src/lint/lint.ts

```diff
@@ -0,0 +1,400 @@
+import type { Node as EcmarkdownNode, Observer } from 'ecmarkdown';
+
+import type { Reporter } from './algorithm-error-reporter-type';
+
+import { parseAlgorithm, visit, emit } from 'ecmarkdown';
+import {
+  Grammar as GrammarFile,
+  SyncHost,
+  EmitFormat,
+  Parser as GrammarParser,
+  SyntaxKind,
+  SymbolSpan,
+  Diagnostics,
+  Production,
+  Parameter,
+  skipTrivia,
+} from 'grammarkdown';
+
+import { getLocation, getProductions, rhsMatches } from './utils';
+import lintAlgorithmLineEndings from './rules/algorithm-line-endings';
+
+function composeObservers(...observers: Observer[]): Observer {
+  return {
+    enter(node: EcmarkdownNode) {
+      for (let observer of observers) {
+        observer.enter?.(node);
+      }
+    },
+    exit(node: EcmarkdownNode) {
+      for (let observer of observers) {
+        observer.exit?.(node);
+      }
+    },
+  };
+}
+
+let algorithmRules = [lintAlgorithmLineEndings];
+
+export type LintingError = { line: number; column: number; message: string };
+
+/*
+Currently this checks
+- grammarkdown's built-in sanity checks
+- the productions in the definition of each early error and SDO are defined in the main grammar
+- those productions do not include `[no LineTerminator here]` restrictions or `[+flag]` gating
+- the algorithm linting rules imported above
+
+There's more to do:
+https://github.com/tc39/ecmarkup/issues/173
+*/
+export function lint(
```

Comment:
Can we break this down into smaller functions and have this function just coordinate the linting results?

Reply:
I am not convinced this is an improvement.

Reply:
Done, though.